javascript/ql/lib/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll

```diff
@@ -109,470 +122,554 @@ module TaintedPath {
+          propertyName = read.getPropertyName() and
+          // example: path.parse('/foo/../bar.txt') returns:
+          // { root: '/', dir: '/foo/..', base: 'bar.txt', ext: '.txt', name: 'bar' }
+          // But we don't trace flow through base, ext, and name properties, just like
+          // we don't trace flow through calls of path.basename.
+          propertyName = ["root", "dir"]
+        )
+      }
+
+      override predicate step(
+        Label::PosixPath inputLabel, DataFlow::Node inputNode, Label::PosixPath outputLabel,
+        DataFlow::Node outputNode
+      ) {
+        input = inputNode and
+        output = outputNode and
+        (
+          propertyName = "root" and outputLabel = inputLabel.toAbsolute().toNormalized()
+          or
+          propertyName = "dir" and outputLabel = inputLabel
+        )
+      }
     }
 
     /**
-     * Gets the input path to be normalized.
+     * A call that preserves taint without changing the flow label.
      */
-    DataFlow::Node getInput() { result = input }
+    class PreservingPathCall extends PathTransformationNode {
+      PreservingPathCall() {
+        this = NodeJSLib::Path::moduleMember(["dirname", "format"]).getACall() and
+        input = this.getAnArgument() and
+        output = this
+        or
+        // non-global replace or replace of something other than /\.\./g, /[/]/g, or /[\.]/g.
+        this instanceof StringReplaceCall and
+        input = this.getReceiver() and
+        output = this and
+        not exists(RegExpLiteral literal, RegExpTerm term |
+          this.(StringReplaceCall).getRegExp().asExpr() = literal and
+          this.(StringReplaceCall).isGlobal() and
+          literal.getRoot() = term
+        |
+          term.getAMatchedString() = "/" or
+          term.getAMatchedString() = "." or
+          term.getAMatchedString() = ".."
+        ) and
+        not this instanceof DotDotSlashPrefixRemovingReplace
+      }
 
-    /**
-     * Gets the normalized path.
-     */
-    DataFlow::Node getOutput() { result = output }
-  }
+      override predicate step(
+        Label::PosixPath inputLabel, DataFlow::Node inputNode, Label::PosixPath outputLabel,
+        DataFlow::Node outputNode
+      ) {
+        inputNode = input and
+        outputNode = output and
+        outputLabel = inputLabel
+      }
+    }
 
-  /**
-   * A call that preserves taint without changing the flow label.
-   */
-  class PreservingPathCall extends DataFlow::CallNode {
-    DataFlow::Node input;
-    DataFlow::Node output;
+    /** A call to the path.join method. */
+    class JoinPathCall extends PathTransformationNode {
+      JoinPathCall() {
+        this = NodeJSLib::Path::moduleMember("join").getACall() and
+        input = this.getAnArgument() and
+        output = this
+      }
 
-    PreservingPathCall() {
-      this =
-        NodeJSLib::Path::moduleMember(["dirname", "toNamespacedPath", "parse", "format"]).getACall() and
```

Comment:
I'm confused by what you've done with this class. 

It seems to me that you've changed the behavior of the first case (pasted below) in this "refactoring" commit.  
```CodeQL
NodeJSLib::Path::moduleMember(["dirname", "toNamespacedPath", "parse", "format"]).getACall() and
input = this.getAnArgument() and
output = this
```

This class was used in the `isPosixPathStep` predicate to just forward a step while preserving the flow-labels.  

It seems to me that you've changed it. 

The `"dirname"` and `"format"` case is in the `PreservingPathCall` class, with the same behavior. 
The `"toNamespacedPath"` case is in `ToNamespacedPathCall`, with the same behavior.  

But the `"parse"` case is in a new class `ParseCall` with different behavior.   
You've added some property checks based on `["root", "dir"]` 

I don't think that check works, but that's another matter. (`CallNode::getALocalUse()` can never be a `PropRead`).

Reply:
Sorry this reply took so long, I got really bogged down with other work and then time off.

What I tried to do here was to track through `parse` more precisely (it gives you an object which contains the `root`, `dir`, `extension`, `baseName` of a path). In principle, we may add more precise tracking of those individual properties — but that's neither a refactoring, nor what this PR really needs to be adding.

As the PR big enough on its own, **I've amend the refactoring PR to remove the class `ParseCall` altogether** (https://github.com/github/codeql/compare/7193fd691a391fb2f27db708b7877c648f797589..8b817fe555a0fdee7caed60ba44be1fc11ac39b7) and moved `parse` back to its friends `dirname` and `format`.

Thank you 👍